File: fizz-buzz/fizz_buzz_csapp_06/FizzBuzz.Tests/RendererTests.cs
namespace FizzBuzz.Tests;

public class RendererTests {
  int displayedCount = 0;
  readonly List<string> displayed = new();

  [Fact]
  public void ShowNumbers_DisplayAllNumbers() {
    Renderer renderer = new(Display);
    renderer.ShowNumbers(2, Render);
    Assert.Equal(2, displayedCount);
  }

  [Fact]
  public void ShowNumbers_RenderNumber() {
    Renderer renderer = new(Display);
    renderer.ShowNumbers(1, Render);
    Assert.Equal("o1", displayed[0]);
  }

  void Display(string text) {
    displayedCount++;
    displayed.Add(text);
  }

  string Render(int number) => $"o{number}";
}

File: fizz-buzz/fizz_buzz_pyapp_06/RendererTests.py
import unittest

from FizzBuzz import fizz_buzz
from Renderer import Renderer


class RendererTestCase(unittest.TestCase):
    def setUp(self):
        self.displayedCount = 0
        self.displayed = []
        self.renderer = Renderer(self.display)

    def display(self, text):
        self.displayedCount += 1
        self.displayed.append(text)

    @staticmethod
    def render(number):
        return f'_x{number}x_'

    def test_show_numbers_should_display_each_number(self):
        self.renderer.show_numbers(2, self.render)
        self.assertEqual(2, self.displayedCount)

    def test_show_numbers_should_render_number(self):
        self.renderer.show_numbers(1, self.render)
        self.assertEqual('_x1x_', self.displayed[0])

    def test_fizz_buzz(self):
        self.renderer = Renderer(print)
        self.renderer.show_numbers(100, fizz_buzz)


if __name__ == '__main__':
    unittest.main()
